### PR TITLE
Sync BLI to OSS

### PR DESCRIPTION
**Pull request by quesurifn**

### CODE REVIEW

File: src/aegis.js [hunk 1]
```diff
@@ -93,11 +93,18 @@ const router = {
...
+                port?.methods?.join('|').toLowerCase() || ''
+
               if (port.path) {
                 routeOverrides.set(port.path, portName)
               }
 
-              if (checkAllowedMethods(ctrl, method)) {
+              if (
+                checkAllowedMethods(ctrl, method) &&
+                (!port.methods ||
```
Comment (quesurifn) [inline src/aegis.js:105]: Vinodh's method fix

File: src/domain/datasource-factory.js [hunk 2]
```diff
@@ -270,7 +274,7 @@ const DataSourceFactory = (() => {
 
     if (adapterName) return adapters[adapterName] || DefaultDataSource
 
-    if (spec?.datasource) {
+    if (spec?.datasource?.factory) {
```
Comment (quesurifn) [inline src/domain/datasource-factory.js:277]: Create indexes

File: src/domain/datasource-factory.js [hunk 3]
```diff
@@ -309,9 +313,12 @@ const DataSourceFactory = (() => {
     const DsClass = createDataSourceClass(spec, options)
     const DsExtendedClass = extendDataSourceClass(DsClass, options)
 
+    if (spec.datasource) {
```
Comment (quesurifn) [inline src/domain/datasource-factory.js:316]: Create indexes

File: src/domain/datasource.js [hunk 4]
```diff
@@ -264,13 +274,6 @@ export default class DataSource {
    */
   async load(options) {}
 
-  /**
-   *
-   */
-  async count() {
-    return this.countSync()
```
Comment (quesurifn) [inline src/domain/datasource.js:271]: Remove dupe count method found via eslint

File: src/domain/use-cases/invoke-port.js [hunk 5]
```diff
@@ -46,27 +46,43 @@ export default function makeInvokePort({
       return threadpool.runJob(invokePort.name, input, modelName)
     } else {
       try {
-        const { id = null, port = null } = input
+        let { id = null, port = null, method, path } = input
```
Comment (quesurifn) [inline src/domain/use-cases/invoke-port.js:49]: Custom paths work

File: src/domain/util/app-error.js [hunk 6]
```diff
@@ -6,6 +6,10 @@
  * @returns
  */
 export function AppError(error, code = 400, cause = null) {
+  if (error.code > 500) {
```
Comment (quesurifn) [inline src/domain/util/app-error.js:9]: Custom indexes work. Duplicate key mongo errors throw a code that's outside of the HTTP spec

File: wasm/assembly/index.ts [hunk 7]
```diff
@@ -2,6 +2,9 @@
 
 @external('env', 'console.log')
 declare function consoleLog (s: string): void
+declare function update (changes: object, runValidation: boolean): void
```
Comment (quesurifn) [inline wasm/assembly/index.ts:5]: ???

No idea what this is. Should we remove?

File: src/adapters/webassembly/wasm-interop.js [hunk 8]
```diff
@@ -221,3 +221,30 @@ exports.WasmInterop = function (wasmExports) {
     callWasmFunction,
   })
 }
+
+/**
+ * Construct an object from the key-value pairs in the multidimensional array
+ * @param {number} ptr - pointer to the address of the array of string arrays
+ * @returns {Readonly<{object}>}
+ */
+// function constructObject (ptr, unpin = true) {
```
Comment (quesurifn) [inline src/adapters/webassembly/wasm-interop.js:230]: Should we remove?

File: src/adapters/datasources/datasource-mongodb.js [hunk 9]
```diff
@@ -211,17 +249,132 @@ export class DataSourceMongoDb extends DataSource {
...
     if (skip) cursor = cursor.skip(skip)
     if (limit) cursor = cursor.limit(limit)
+
     return cursor
   }
 
-  processOptions(param) {
-    const { options = {}, query = {} } = param
-    return { ...options, ...processQuery(query) }
+  /**
```
Comment (quesurifn) [inline src/adapters/datasources/datasource-mongodb.js:263]: This requires a good look.

File: src/adapters/datasources/datasource-mongodb.js [hunk 10]
```diff
@@ -98,23 +107,49 @@ export class DataSourceMongoDb extends DataSource {
...
   }
 
   async collection() {
-    try {
-      return (await this.connection()).db(this.namespace).collection(this.name)
-    } catch {}
+    return (await this.connection()).db(this.namespace).collection(this.name)
+  }
+
+  async createIndexes(client) {
```
Comment (quesurifn) [inline src/adapters/datasources/datasource-mongodb.js:133]: Create index work.

File: src/adapters/datasources/datasource-mongodb.js [hunk 11]
```diff
@@ -98,23 +107,49 @@ export class DataSourceMongoDb extends DataSource {
...
       return client
     } catch (error) {
       console.error({ fn: this.connection.name, error })
     }
   }
 
   async collection() {
-    try {
-      return (await this.connection()).db(this.namespace).collection(this.name)
-    } catch {}
```
Comment (quesurifn) [inline src/adapters/datasources/datasource-mongodb.js:110]: Squelching error

File: src/adapters/datasources/datasource-mongodb.js [hunk 12]
```diff
@@ -1,22 +1,30 @@
...
+    }, // reference BSON datatypes https://www.mongodb.com/docs/manual/reference/bson-types/
+  },
+})
 
 const url = process.env.MONGODB_URL || 'mongodb://localhost:27017'
-const configRoot = require('../../config').hostConfig
+
 const dsOptions = configRoot.adapters.datasources.DataSourceMongoDb.options || {
-  runOffline: true,
+  runOffline: false,
```
Comment (quesurifn) [inline src/adapters/datasources/datasource-mongodb.js:24]: So we can test properly.

File: src/adapters/datasources/datasource-mongodb.js [hunk 13]
```diff
@@ -1,22 +1,30 @@
...
+import DataSource from '../../domain/datasource'
 
 const HIGHWATERMARK = 50
 
-const mongodb = require('mongodb')
-const { MongoClient } = mongodb
-const { Transform, Writable } = require('stream')
-const qpm = require('query-params-mongo')
-const processQuery = qpm()
+const processQuery = qpm({
```
Comment (quesurifn) [inline src/adapters/datasources/datasource-mongodb.js:12]: Null value in mongo fix by @vi-ssc

File: src/adapters/controllers/post-model.js [hunk 14]
```diff
@@ -21,7 +21,10 @@ export default function postModelFactory(createModel) {
           'Last-Modified': new Date().toISOString(),
         },
         statusCode: 201,
-        body: { modelId: model.id },
+        body: {
+          status: 'ok',
```
Comment (quesurifn) [inline src/adapters/controllers/post-model.js:25]: Consistent error messages for Prerna

File: src/adapters/controllers/post-invoke-port.js [hunk 15]
```diff
@@ -13,6 +13,10 @@ export default function anyInvokePortFactory(invokePort) {
       const result = await invokePort({
         port: httpRequest.params.port,
         args: httpRequest.body,
+        params: httpRequest.query,
+        method: httpRequest.method,
```
Comment (quesurifn) [inline src/adapters/controllers/post-invoke-port.js:17]: @vi-ssc (or @velosomarcus ?) fix for methods, params and headers in custom ports.

File: src/adapters/controllers/patch-model.js [hunk 16]
```diff
@@ -21,7 +21,10 @@ export default function patchModelFactory(editModel) {
           'Last-Modified': new Date().toISOString(),
         },
         statusCode: 201,
-        body: { modelId: model.id },
+        body: {
```
Comment (quesurifn) [inline src/adapters/controllers/patch-model.js:24]: Consistent return value formatting fix for Prerna

File: src/adapters/controllers/delete-model.js [hunk 17]
```diff
@@ -15,7 +15,10 @@ export default function deleteModelFactory(removeModel) {
           'Last-Modified': new Date().toISOString(),
         },
         statusCode: 201,
-        body: { modelId: model.id },
+        body: {
+          status: 'ok',
```
Comment (quesurifn) [inline src/adapters/controllers/delete-model.js:19]: Consistent return value formatting fix for Prerna

File: .eslintrc.js [hunk 18]
```diff
@@ -1,12 +1,19 @@
...
   },
   extends: ['standard', 'plugin:prettier/recommended'],
   parserOptions: {
     ecmaVersion: 11,
     sourceType: 'module',
+    ecmaFeatures: {
+      modules: true,
+    },
   },
+  extends: ['eslint:recommended', 'plugin:prettier/recommended'],
```
Comment (quesurifn) [inline .eslintrc.js:17]: Add ESLINT recommended work as tracked in Monday

File: .eslintrc.js [hunk 19]
```diff
@@ -1,12 +1,19 @@
 module.exports = {
   env: {
     browser: true,
+    es6: true,
+    node: true,
```
Comment (quesurifn) [inline .eslintrc.js:5]: Allows linter to be aware of things like 

```
process.on('SIGTERM', () => {

})
```

**Review by tysonrm — APPROVED** [review]

**Pull request merged**